### PR TITLE
fix for #906, 1) windows paths like C:\a\b weren't accepted because b…

### PR DESCRIPTION
**Pull request by downpoured**
1) windows paths like C:\a\b weren't accepted because backslashes were treated as escape sequences, 
2) common paths like C:\Program Files\Foo\Foo.exe weren't accepted because of the space in the path

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks for the patch. Do you know what part of the code was treating "\" as an escape character? Maybe, rather than escaping the path, whatever was incorrectly treating \ as escape character should be fixed?

**Comment by downpoured** [comment]
It's splitCommandString() that is both treating \ as an escape character and treating spaces as separating arguments. This is right for Linux, where paths don't usually have spaces, but isn't expected in Windows.

True, another approach could be to simply skip splitCommandString() on Windows and pass the string as one argument. I doubt that Windows users would need to pass flags to a .exe. This would be a simpler ~5 line patch, along with a new string, since the documentation string says "The editor command (may include arguments) that will be used..." would have a Windows version that says "The editor command that will be used..."

**Comment by downpoured** [comment]
t's splitCommandString() that is both treating \ as an escape character and treating spaces as separating arguments. This is right for Linux, where paths don't usually have spaces, but isn't expected in Windows.

True, a simpler approach could be to simply skip splitCommandString() on Windows and pass the string as one argument. I doubt that Windows users would need to pass flags to a .exe. This would be a simpler ~5 line patch, along with a new string, since the documentation string says "The editor command (may include arguments) that will be used..." would have a Windows version that says "The editor command that will be used..."

If that would be preferred, I'll update the pr.

**Comment by downpoured** [comment]
It's splitCommandString() that is both treating \ as an escape character and treating spaces as separating arguments. This is right for Linux, where paths don't usually have spaces, but isn't expected in Windows.

True, a simpler approach could be to simply skip splitCommandString() on Windows and pass the string as one argument. I doubt that Windows users would need to pass flags to a .exe. This would be a simpler ~5 line patch, along with a new string, since the documentation string says "The editor command (may include arguments) that will be used..." would have a Windows version that says "The editor command that will be used..."

If that would be preferred, I'll update the pr.

**Comment by laurent22** [comment]
> It's splitCommandString() that is both treating \ as an escape character and treating spaces as separating arguments. This is right for Linux, where paths don't usually have spaces, but isn't expected in Windows.

Hmm, yes I think the bug is indeed in splitCommandString(). I've converted this function from some of my Golang code but it was used in a different context where escaping needed to be supported. However in this case we don't ever need escaping - the app should just take whatever string the user provided and run this.

So yes I think if you could fix this function by making its ability to escape option it should be fine. Like the signature would be this:

```
function splitCommandString(command, options =null) {
    options = {...options};
    if (!('handleEscape' in options)) options.handleEscape = true;
    /// etc.
}
```

For the spaces in Windows, the splitCommandString function supports quotes around paths so we don't need to do something special for this. Perhaps we could add a description below the field though - something like "If you path contains spaces, add double quotes around it".

**Comment by downpoured** [comment]
When I update joplin.pot, which script should I run to regenerate locales/*.js ? (I'm guessing something related to gettext)

**Comment by laurent22** [comment]
You don't need to regenerate the .pot file, I take care of it when I rebuild all the translations. If you want a string to be translated, simply wrap it in `_()`.

File: ReactNativeClient/lib/string-utils.js [hunk 1]
```diff
@@ -122,7 +122,26 @@ function wrap(text, indent, width) {
...
+async function prepareCmdStringWindows(cmd, fnExists) {
+	// in windows, paths often contain spaces, so
+	// if there are spaces and the entire string points to an existing file,
+	// it's probably something like C:\Program Files\Foo\Foo.exe
+	if (cmd.indexOf(" ") != -1 && cmd.indexOf('"') == -1 && await fnExists(cmd)) {
+		cmd = '"' + cmd + '"';
+	}
+
+	// in windows, \ usually is a path delim, not an escape sequence
+	cmd = cmd.replace(/\\/g, '\\\\');
```
Comment (laurent22) [inline ReactNativeClient/lib/string-utils.js]: Is there any reason we still need this, since in splitCommandString you are now disabling escape?

**Comment by laurent22** [comment]
Sorry I missed your new commits from a few days ago. I think it's almost ready to merge but I just have one question on one of the changes. If you could just confirm on this that would be great.

File: ReactNativeClient/lib/string-utils.js (hunk 1 above)
Comment (downpoured) [reply ReactNativeClient/lib/string-utils.js]: Thanks, good catch.

**Comment by downpoured** [comment]
Sounds good.

**Comment by laurent22** [comment]
Thanks @downpoured!

**Comment by downpoured** [comment]
Thanks!

By the way, I'm thinking of changing my username on here to 'moltenjs' within the next few days, GitHub seems to handle this fine, but letting you know just in case.

**Pull request merged**